### PR TITLE
Optimize Softmax with online normalizer algorithm

### PR DESCRIPTION
**Pull request by daemyung**
## 🙏 Describe the pull request

- Online Softmax를 구현하여 성능을 개선 했습니다.

## 💬 Additional context

- Triton에 문제가 있어서 0차원 Softmax는 지원하지 못합니다.
- 텐서가 큰 경우, 메모리 대역폭에 의해 성능이 10%이상 향상됩니다.

plot

## ✅ Checklist

- [x] Code follows the project's coding conventions and style.
- [x] Tests have been added or updated to cover the changes.
- [x] Documentation has been updated, if necessary.

### CODE REVIEW

File: trident/module.py [hunk 1]
```diff
@@ -713,7 +713,7 @@ def forward(self, input):
 
 
 class Softmax(torch.nn.Module):
-    def __init__(self, dim=None):
+    def __init__(self, dim: int = None):
```
Comment (mejai1206) [inline trident/module.py:716]: dim의 기본값으로 1이 있는 것은 어떨까요?

Comment (daemyung) [reply trident/module.py:716]: PyTorch의 경우 기본값이 `None`입니다.

**Review by mejai1206 — APPROVED** [review]
LGTM

**Review by kakao-steve-ai — APPROVED** [review]
코드는 좋아보입니다. dim=0은 triton 문제로 지원되지 않는다고 하셨는데, 왜 그런건가요?

**Comment by daemyung** [comment]
> 코드는 좋아보입니다. dim=0은 triton 문제로 지원되지 않는다고 하셨는데, 왜 그런건가요?

Triton에서 계산 결과가 이상합니다. Triton에 이슈로 만들어서 리포트할 예정입니다.

**Comment by kakao-steve-ai** [comment]
> > 코드는 좋아보입니다. dim=0은 triton 문제로 지원되지 않는다고 하셨는데, 왜 그런건가요?
> 
> Triton에서 계산 결과가 이상합니다. Triton에 이슈로 만들어서 리포트할 예정입니다.

네 알겠습니다.

**Comment by daemyung** [comment]
> > > 코드는 좋아보입니다. dim=0은 triton 문제로 지원되지 않는다고 하셨는데, 왜 그런건가요?
> > 
> > 
> > Triton에서 계산 결과가 이상합니다. Triton에 이슈로 만들어서 리포트할 예정입니다.
> 
> 네 알겠습니다.

리포트하려고 재현을 했더니,,, 잘 되네요? 역전파 메모리 사용량 최적화 하면서 다른 축도 받을 수 있도록 수정하겠습니다.
